### PR TITLE
MM-10264: Adds system scheme to permissions import/export.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary
Adds system scheme to permissions import/export.

#### Ticket Link
[MM-10264](https://mattermost.atlassian.net/browse/MM-10264)

#### Checklist
- [ ] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: app/permissions.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 const permissionsExportBatchSize = 100
+const systemSchemeName = "9CE2A3E8-D8EA-4621-BCC0-27DED244D37B" // Prevents collisions with user-created schemes.
```
Comment (jespino) [inline app/permissions.go]: If we going to include a "Magic number" to define the systemSchemeName, I would prefer something like `00000000-0000-0000-0000-000000000000`, will avoid the collisions and looks like something fixed.

File: app/permissions.go [hunk 2]
```diff
@@ -101,6 +102,31 @@ func (a *App) ExportPermissions(w io.Writer) error {
 
 	}
 
+	defaultRoleNames := []string{}
+	for _, dr := range model.MakeDefaultRoles() {
+		defaultRoleNames = append(defaultRoleNames, dr.Name)
+	}
+
```
Comment (jespino) [inline app/permissions.go]: I'm seeing the code and it's a bit overkill to generate de default roles to get the names. Actually for the default roles the ID is equal to the Name, so I would change the model/role.go to declare a constant DEFAULT_ROLE_NAMES with the list of default roles, and in the function MakeDefaultRoles I would use the same constant for the map key and for the Name field value.

Comment (jespino) [reply app/permissions.go]: Something like

```go
const DEFAULT_ROLE_NAMES = []string{
    SYSTEM_USER_ROLE_ID,
    SYSTEM_ADMIN_ROLE_ID,
    SYSTEM_POST_ALL_ROLE_ID,
    SYSTEM_POST_ALL_PUBLIC_ROLE_ID,
    SYSTEM_USER_ACCESS_TOKEN_ROLE_ID,
    TEAM_USER_ROLE_ID,
    TEAM_ADMIN_ROLE_ID,
    TEAM_POST_ALL_ROLE_ID,
    TEAM_POST_ALL_PUBLIC_ROLE_ID,
    CHANNEL_USER_ROLE_ID,
    CHANNEL_ADMIN_ROLE_ID,
}
```

File: app/permissions.go [hunk 3]
```diff
@@ -101,6 +102,31 @@ func (a *App) ExportPermissions(w io.Writer) error {
...
+	schemeExport, err := json.Marshal(&model.SchemeConveyor{
+		Name:  systemSchemeName,
+		Roles: roles,
+	})
+	if err != nil {
+		return err
+	}
+
+	schemeExport = append(schemeExport, []byte("\n")...)
```
Comment (jespino) [inline app/permissions.go]: I think it would be enough simply `schemeExport = append(schemeExport, '\n')`

File: app/permissions.go [hunk 4]
```diff
@@ -116,6 +142,25 @@ func (a *App) ImportPermissions(jsonl io.Reader) error {
...
+			for _, roleIn := range schemeConveyor.Roles {
+				dbRole, err := a.GetRoleByName(roleIn.Name)
+				if err != nil {
+					return errors.New(err.Message)
+				}
+				_, err = a.PatchRole(dbRole, &model.RolePatch{
+					Permissions: &roleIn.Permissions,
+				})
+				if err != nil {
+					for _, schemeID := range createdSchemeIDs {
```
Comment (jespino) [inline app/permissions.go]: I think this clean up must be called for any return sentence returning an error, and some of the errors return don't clean up the created schemes.

Comment (mkraft) [reply app/permissions.go]: You're right it's safer.

**Comment by mkraft** [comment]
@jespino I have added the rollback to all error cases. 

As for the two suggestions: I agree with them and if you're OK with it I'll do them separately in the near future given that they're non-essential refactors and code-complete was yesterday. How does that sound?

**Comment by jespino** [comment]
Sounds good :)

**Pull request merged**